library/src/main/java/com/bumptech/glide/load/resource/bitmap/GlideBitmapFactoryWrapper.java

```diff
@@ -0,0 +1,291 @@
+     * <p>If the input gainmap is not single channel, then this method will just return the original
+     * gainmap.
+     */
+    public static Gainmap convertSingleChannelGainmapToTripleChannelGainmap(Gainmap gainmap) {
+      Bitmap gainmapContents = gainmap.getGainmapContents();
+      if (gainmapContents.getConfig() != Config.ALPHA_8) {
+        return gainmap;
+      }
+      Bitmap newContents = gainmapContents.copy(Config.ARGB_8888, /* isMutable= */ false);
+      Gainmap newGainmap = new Gainmap(newContents);
+      float[] tempFloatArray = gainmap.getRatioMin();
+      newGainmap.setRatioMin(tempFloatArray[0], tempFloatArray[1], tempFloatArray[2]);
+      tempFloatArray = gainmap.getRatioMax();
+      newGainmap.setRatioMax(tempFloatArray[0], tempFloatArray[1], tempFloatArray[2]);
+      tempFloatArray = gainmap.getGamma();
+      newGainmap.setGamma(tempFloatArray[0], tempFloatArray[1], tempFloatArray[2]);
+      tempFloatArray = gainmap.getEpsilonSdr();
+      newGainmap.setEpsilonSdr(tempFloatArray[0], tempFloatArray[1], tempFloatArray[2]);
+      tempFloatArray = gainmap.getEpsilonHdr();
+      newGainmap.setEpsilonHdr(tempFloatArray[0], tempFloatArray[1], tempFloatArray[2]);
+      newGainmap.setDisplayRatioForFullHdr(gainmap.getDisplayRatioForFullHdr());
+      newGainmap.setMinDisplayRatioForHdrTransition(gainmap.getMinDisplayRatioForHdrTransition());
+      return newGainmap;
+    }
+  }
+
+  /**
+   * Determines if a gainmap decoding workaround is required to mitigate an Android U bug with
+   * decoding bitmaps with gainmaps. When the following conditions are present, Android U will not be
+   * able to decode a gainmap, with hardware bitmap operation failing for the gainmap:
+   *
+   * <ul>
+   *   <li>The HWUI is configured to use skiagl.
+   *   <li>The gainmap is single channel.
+   *   <li>The bitmap owning the bitmap is a hardware bitmap.
+   * </ul>
+   *
+   * <p>Callers should use this class to determine whether to apply a workaround, e.g., modifying the
+   * gainmap to be triple channel and software decode it.
+   */
+  public static final class GainmapDecoderWorkaroundStateCalculator {
+    private static final String TAG = "GainmapWorkaroundCalc";
+
+    /** Meomizes result of test to see if the device is susceptible to the gainmap decoding bug. */
+    @GuardedBy("GainmapDecoderWorkaroundStateCalculator.class")
+    private static final AtomicReference<Boolean> REQUIRES_GAIN_MAP_FIX = new AtomicReference<>();
+
+    private GainmapDecoderWorkaroundStateCalculator() {}
+
+    /**
+     * Returns true if a gainmap decoding workaround is required to mitigate an Android U bug. This
+     * method tests for the presence of the bug, which only affects hardware bitmaps, and caches the
+     * result in memory.
+     *
+     * <p>This method is thread-safe.
+     *
+     * @param options which will be used to decode the gainmap.
+     */
+    static synchronized boolean needsGainmapDecodeWorkaround(Options options) {
```

Comment:
I think you can be a bit more efficient with the locking, some of the preamble here doesn't have to be synchronized, and they don't need a shared lock until REQUIRES_GAIN_MAP_FIX is found to be null?

Not sure if it matters that much, but we do know these methods are invoked from 4 threads concurrently.

Reply:
Done. Sorry this had violated the Effective Java rule of avoid excessive synchronization. 

I switched over to using GlideSupplier, which is the Glide way of memoizing expensive compuatiohns.